projects/swimlane/ngx-ui/src/lib/components/stepper/stepper.component.ts

```diff
@@ -0,0 +1,203 @@
+import {
+  Component,
+  ChangeDetectionStrategy,
+  ViewEncapsulation,
+  ChangeDetectorRef,
+  Input,
+  Output,
+  EventEmitter,
+  ContentChildren,
+  QueryList,
+  OnDestroy
+} from '@angular/core';
+import { coerceNumberProperty, coerceBooleanProperty } from '@angular/cdk/coercion';
+import { Subject } from 'rxjs';
+import { takeUntil } from 'rxjs/operators';
+
+import { StepperPosition } from './stepper-position.enum';
+import { StepComponent } from './step.component';
+import { stepperAnimations } from './stepper.animation';
+import { StepperAnimationStates } from './stepper-animation-states.enum';
+import { StepperBarAnimationStates } from './stepper-bar-animation-states.enum';
+
+@Component({
+  exportAs: 'ngxStepper',
+  selector: 'ngx-stepper',
+  templateUrl: './stepper.component.html',
+  styleUrls: ['./stepper.component.scss'],
+  host: {
+    class: 'ngx-stepper',
+    '[class.ngx-stepper--readonly]': 'readonly',
+    '[class.ngx-stepper--top]': 'position === StepperPosition.Top',
+    '[class.ngx-stepper--bottom]': 'position === StepperPosition.Bottom',
+    '[class.ngx-stepper--left]': 'position === StepperPosition.Left',
+    '[class.ngx-stepper--right]': 'position === StepperPosition.Right'
+  },
+  animations: [
+    stepperAnimations.horizontalStepTransition,
+    stepperAnimations.verticalStepTransition,
+    stepperAnimations.horizontalBarTransition,
+    stepperAnimations.verticalBarTransition
+  ],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+  encapsulation: ViewEncapsulation.None
+})
+export class StepperComponent implements OnDestroy {
+  @Input() position = StepperPosition.Top;
+
+  @Input()
+  get completeIcon() {
+    return this._completeIcon;
+  }
+  set completeIcon(v: string) {
+    if (this._steps) {
+      for (const step of this._steps) {
+        if (!step.completeIcon || step.completeIcon === this._completeIcon) {
+          step.completeIcon = v;
+        }
+      }
+    }
+
+    this._completeIcon = v;
+    this._cdr.markForCheck();
+  }
+
+  @Input()
+  get active() {
+    return this._active;
+  }
+  set active(v: number) {
+    v = coerceNumberProperty(v);
+
+    if (v !== undefined && !isNaN(v) && v !== this._active && v >= 0 && (!this._steps || v < this._steps.length)) {
+      this._active = v;
+
+      if (this._steps) {
+        for (const step of this._steps) {
+          step.active = v;
+        }
+      }
+
+      this.activeChange.emit(this._active);
+      this._cdr.markForCheck();
+    }
+  }
+
+  @Input()
+  get readonly() {
+    return this._readonly;
+  }
+  set readonly(v: boolean) {
+    this._readonly = coerceBooleanProperty(v);
+    this._cdr.markForCheck();
+  }
+
+  @Output() activeChange = new EventEmitter<number>();
+
+  @ContentChildren(StepComponent)
+  get steps() {
+    return this._steps;
+  }
+  set steps(v) {
+    this._steps = v;
+    this._destroy$.next();
+
+    for (const item of this._steps.map((step, i) => ({ step, i }))) {
+      setTimeout(() => {
```

Comment:
so even though all these properties have mark for check, you still needed a setTimeout?

Reply:
Yes, if not you will get weird change detection behavior, this is so that we can make all changes outside angular then do a check after the fact. Not ideal but the only performant solution I could find.

Reply:
figured something like that was the reason, just wanted to double check 👍